packages/upload-client/src/sharding.js

```diff
@@ -79,10 +90,26 @@ export class ShardingStream extends TransformStream {
                 `block will cause CAR to exceed shard size: ${block.cid}`
               )
           }
-          controller.enqueue(await encode(blocks))
-          controller.enqueue(await encode(overflowBlocks, rootCID))
+          controller.enqueue(await encodeCAR(blocks, slices))
+
+          // Finally, re-calc block positions from blocks we moved out of the
+          // CAR that was too big.
+          overflowCurrentLength = 0
+          /** @type {Map<import('./types.js').SliceDigest, import('./types.js').Position>} */
+          const overflowSlices = new DigestMap()
+          for (const block of blocks) {
+            const overflowBlockHeaderLength = blockHeaderEncodingLength(block)
+            overflowSlices.set(block.cid.multihash, [headerLength + overflowCurrentLength + overflowBlockHeaderLength, block.bytes.length])
+            overflowCurrentLength += overflowBlockHeaderLength + block.bytes.length
+          }
+          controller.enqueue(await encodeCAR(overflowBlocks, overflowSlices, rootCID))
         } else {
-          controller.enqueue(await encode(blocks, rootCID))
+          // adjust offsets for longer header in final shard
+          const diff = headerLength - headerEncodingLength()
+          for (const slice of slices.values()) {
+            slice[0] += diff
+          }
+          controller.enqueue(await encodeCAR(blocks, slices, rootCID))
```

Comment:
Putting a root CID in the last CAR is a massive PITA. I wish we didn't do that.

Reply:
Wait, I thought we were putting it in the first shard.

Let's put it as a separate thing that is not in the set of shards. It would be easier to change that now.

Reply:
Yes I would like all shards to be rootless and track the DAG root in `upload/add` and/or content claims.